### ISSUE TITLE
How to implement a loss function with multi output?

### ISSUE THREAD

**Issue by cgcodeboy** [keras-team/keras#13453]
## I have a problem when I want to implement a loss function with two output in keras. The model just like this:
```
input_tensor = Input(shape = my_shape)
...... # some customer code
output_1 = Dense(1, activation = 'sigmoid')(vector_1)
......# some customer code
output_2 = Dense(1, activation = 'sigmoid')(vector_2)

model = Model(input_tensor,[output_1, output_2])
```
Now, I want to implement a loss function which can calculate three different loss, the first one is the binary cross entropy of output_1, the second is the binary cross entropy of output_2, the last  one is the MSE between output_1 and output_2, and I want to integrate the three loss into only one loss function，how can I implement it? I hope you can give me a good suggestion~

**Comment by maxhardt** [comment]
Using the functional model API, multi head losses are usually quite straightforward to implement:
```
# note that I added the names of the layer for loss reference later 
output_1 = Dense(1, activation = 'sigmoid', name="output_1")(vector_1)
output_2 = Dense(1, activation = 'sigmoid', name="output_2")(vector_2)
model = Model(input_tensor,[output_1, output_2])

losses = {
    "output_1": "binary_cross_entropy",
    "output_2": "binary_cross_entropy"
}
loss_weights = {
    "output_1": 1.0,
    "output_2": 1.0
}
model.compile(loss=losses, loss_weights=loss_weights)
```
However, It gets tricky when defining a custom loss function that does not follow the keras norm with `def custom_loss(y_true, y_pred)`. However, you could get away with slicing the predictions in a custom loss function as follows:

```
def custom_loss(y_true, y_pred):
    loss_output_1 = keras.losses.binary_crossentropy(y_pred[:, 0], y_true)
    loss_output_2 = keras.losses.binary_crossentropy(y_pred[:, 1], y_true)
    mse_loss = keras.losses.mean_squared_error(y_pred[:, 0], y_pred[:, 1])
    return loss_output_1 + loss_output_2 + mse_loss

model.compile(loss=custom_loss)
```

**Comment by maxhardt** [comment]
also interesting #2121

**Comment by falk-stefan** [comment]
@maxhardt thanks for the example above. But one question: What if I wanted to compute only one function - one combined loss like.

Basically something like this:

```python
class CombinedLoss(Loss):

  def __init__(self, weights):
    self.weights = weights

  def call(y_true, y_pred):
     y1, y2 = y_true
     out1, out2, = y_pred
     w1, w2 = self.weights
     return sum(w2 * mse(y1 - out1), w2 * mse(y2 - out2))


model = Model(inputs=[in_1, in_2], outputs=[out_1, out_2])
model.compile(loss=CombinedLoss(weights=[0.2, 0.8]))
```

I can see a workaround which would look like this:

```python
def dumm_loss(*args, **kwargs):
  return 0.0

# From your example
combined_outputs = concatenate([out_1, out_2], name="combined_output")
model = Model(inputs=[in_1, in_2], outputs=[out_1, out_2, combined_outputs])
model.compile(loss=[dummy_loss, dummy_loss, CombinedLoss(weights=[0.2, 0.8])])
``` 

but is this the best we can do?
